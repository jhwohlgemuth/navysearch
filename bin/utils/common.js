"use strict";var _interopRequireDefault=require("@babel/runtime/helpers/interopRequireDefault");var _defineProperty2=_interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));require("core-js/modules/es.array.iterator"),require("core-js/modules/es.array.sort"),require("core-js/modules/es.object.entries"),require("core-js/modules/es.object.get-own-property-descriptors"),require("core-js/modules/es.string.split");function ownKeys(a,b){var c=Object.keys(a);if(Object.getOwnPropertySymbols){var d=Object.getOwnPropertySymbols(a);b&&(d=d.filter(function(b){return Object.getOwnPropertyDescriptor(a,b).enumerable})),c.push.apply(c,d)}return c}function _objectSpread(a){for(var b,c=1;c<arguments.length;c++)b=null==arguments[c]?{}:arguments[c],c%2?ownKeys(Object(b),!0).forEach(function(c){(0,_defineProperty2.default)(a,c,b[c])}):Object.getOwnPropertyDescriptors?Object.defineProperties(a,Object.getOwnPropertyDescriptors(b)):ownKeys(Object(b)).forEach(function(c){Object.defineProperty(a,c,Object.getOwnPropertyDescriptor(b,c))});return a}const{extname}=require("path"),{invertObj,takeWhile}=require("ramda"),NPC_DOMAIN="http://www.public.navy.mil",MAX_MESSAGE_TEXT_LENGTH=5e3,MESSAGE_TYPE_LOOKUP={NAV:"NAVADMIN",ALN:"ALNAV"},YEAR_FORMAT_LENGTH=2,createMessageId=({type:a,year:b,num:c})=>`${a}${b}${c}`,createNpcPageUrl=({type:a,year:b})=>{// eslint-disable-line no-magic-numbers
return`${NPC_DOMAIN}/bupers-npc/reference/messages/${a}S/Pages/${a}20${(a=>(a+"").length===4?+a.substr(-2):a)(b)}.aspx`},createMessageUrl=({num:a,type:b,year:c})=>{const d=invertObj(MESSAGE_TYPE_LOOKUP)[b],e=parseInt(c,10),f="NAV"===d,g=18>e,h=`bupers-npc/reference/messages/Documents${f&&g?"2":""}`;return`${NPC_DOMAIN}/${h}/${f&&!g||"ALN"===d?`${b}S/`:""}${d}20${c}/${d}${c}${a}.txt`},createYearsString=a=>a.sort().slice(0,-1).map(a=>`20${a}`).join(", ").concat(`${2<a.length?",":""}${1<a.length?" and ":""}20${a.slice(-1)}`),dict=a=>new Map(Object.entries(a)),getCurrentYear=()=>+(new Date().getFullYear()+"").substring(2),isNotNumber=a=>isNaN(+a),isNumberLike=a=>!isNaN(+a),parseMessageName=a=>{const b=takeWhile(isNotNumber,[...a]).join("").toUpperCase(),c=MESSAGE_TYPE_LOOKUP[b],d=a.substring(b.length,b.length+YEAR_FORMAT_LENGTH),e=a.substring(b.length+YEAR_FORMAT_LENGTH),f=createMessageUrl({num:e,type:c,year:d}),g=createMessageId({num:e,type:c,year:d});return{code:b,id:g,num:e,type:c,url:f,year:d}},parseMessageId=a=>{const b=takeWhile(isNotNumber,[...a]).join("").toUpperCase(),c=invertObj(MESSAGE_TYPE_LOOKUP)[b],d=a.substring(b.length,b.length+YEAR_FORMAT_LENGTH),e=a.substring(b.length+YEAR_FORMAT_LENGTH);return parseMessageName(`${c}${d}${e}`)},parseMessageUri=a=>{const[b]=a.split("/").reverse(),[c]=b.split("."),d=extname(a);return _objectSpread({},parseMessageName(c),{ext:d})},partitionByKeyLength=(a,b,c)=>{const d=c[a]+"",e=Math.ceil(d.length/b);return[..."x".repeat(e)].map((e,f)=>{const g=f*b;return _objectSpread({},c,{[a]:d.substring(g,g+b)})})};module.exports={MAX_MESSAGE_TEXT_LENGTH,NPC_DOMAIN,createMessageId,createMessageUrl,createNpcPageUrl,createYearsString,dict,getCurrentYear,isNotNumber,isNumberLike,parseMessageId,parseMessageName,parseMessageUri,partitionByKeyLength};