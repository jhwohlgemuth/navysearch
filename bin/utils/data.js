"use strict";var _interopRequireDefault=require("@babel/runtime/helpers/interopRequireDefault"),_toConsumableArray2=_interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray")),_slicedToArray2=_interopRequireDefault(require("@babel/runtime/helpers/slicedToArray")),_defineProperty2=_interopRequireDefault(require("@babel/runtime/helpers/defineProperty")),_regenerator=_interopRequireDefault(require("@babel/runtime/regenerator")),_asyncToGenerator2=_interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));function ownKeys(a,b){var c=Object.keys(a);if(Object.getOwnPropertySymbols){var d=Object.getOwnPropertySymbols(a);b&&(d=d.filter(function(b){return Object.getOwnPropertyDescriptor(a,b).enumerable})),c.push.apply(c,d)}return c}function _objectSpread(a){for(var b,c=1;c<arguments.length;c++)b=null==arguments[c]?{}:arguments[c],c%2?ownKeys(Object(b),!0).forEach(function(c){(0,_defineProperty2["default"])(a,c,b[c])}):Object.getOwnPropertyDescriptors?Object.defineProperties(a,Object.getOwnPropertyDescriptors(b)):ownKeys(Object(b)).forEach(function(c){Object.defineProperty(a,c,Object.getOwnPropertyDescriptor(b,c))});return a}var _require=require("util"),promisify=_require.promisify,log=require("npmlog"),chalk=require("chalk"),ora=require("ora"),Xray=require("x-ray"),algoliasearch=require("algoliasearch"),axios=require("axios")["default"],_require2=require("ramda"),difference=_require2.difference,_require3=require("./common.js"),MAX_MESSAGE_TEXT_LENGTH=_require3.MAX_MESSAGE_TEXT_LENGTH,createMessageId=_require3.createMessageId,createNpcPageUrl=_require3.createNpcPageUrl,createYearsString=_require3.createYearsString,getCurrentYear=_require3.getCurrentYear,parseMessageId=_require3.parseMessageId,parseMessageUri=_require3.parseMessageUri,partitionByKeyLength=_require3.partitionByKeyLength,bold=chalk.bold,cyan=chalk.cyan,green=chalk.green,scrapeMessageItems=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(b,c){var d,e,f,g,h;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return d=new Xray,e=createNpcPageUrl({type:b,year:c}),a.next=4,axios.get(e);case 4:return f=a.sent,g=f.data,a.next=8,promisify(d(g,"a",[{href:"@href"}]));case 8:return h=a.sent,a.abrupt("return",h.map(function(a){var b=a.href;return b}).filter(function(a){return /[.]txt$/.test(a)}).map(parseMessageUri).filter(function(a){var b=a.code,c=a.num;return[b,c].every(function(a){var b=a.length;return 0<b})}));case 10:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}(),getItem=function(a){var b=a.code,c=a.num,d=a.type,e=a.year,f=a.url,g={code:b,num:c,type:d,url:f,year:e},h=createMessageId(g);return axios.get(f).then(function(a){var b=a.data;return _objectSpread({},g,{id:h,text:b,objectID:h})})["catch"](function(){return _objectSpread({},g,{id:h,objectID:h})})},getItems=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(b,c){var d,e,f,g,h;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return d=function(a){var b=(0,_slicedToArray2["default"])(a,1),d=b[0].type,e=createYearsString(c),f="Fetching ".concat(bold(d)," data for ").concat(bold(e)," (").concat(a.length," items)\n\n");return cyan(f)},e=function(a){var b=a.filter(function(a){var b=a.text;return"undefined"!=typeof b});return"".concat(green("COMPLETE")," ~ ").concat(bold(b.length)," of ").concat(bold(a.length)," items processed\n")},f=function(a){return Promise.all(a.map(getItem))},a.next=5,Promise.all(c.map(function(a){return scrapeMessageItems(b,a)}));case 5:return g=a.sent.flat(1/0),h=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(){var b,c,h,i,j,k,l;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return b=ora(d(g)).start(),a.next=3,f(g);case 3:return c=a.sent,h=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(c){var d;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return d=c.filter(function(a){var b=a.text;return"undefined"==typeof b}),b.text=cyan("Trying to fetch data for ".concat(bold(d.length)," of ").concat(g.length," items")),a.abrupt("return",0===d.length?Promise.resolve([]):f(d));case 3:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}(),a.next=7,h(c);case 7:return i=a.sent,a.next=10,h(i);case 10:return j=a.sent,a.next=13,h(j);case 13:return k=a.sent,l=[].concat((0,_toConsumableArray2["default"])(c),(0,_toConsumableArray2["default"])(i),(0,_toConsumableArray2["default"])(j),(0,_toConsumableArray2["default"])(k)).filter(function(a){var b=a.text;return"undefined"!=typeof b}),b.succeed(e(l)),a.abrupt("return",l);case 17:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}(),a.abrupt("return",0<g.length?h():[]);case 8:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}(),getSavedItems=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(b){var c,d,e,f,g,h,i,j;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return c=b.id,d=b.key,e=b.name,f=void 0===e?"message":e,g=algoliasearch(c,d),h=g.initIndex(f),a.next=5,h.setSettings({paginationLimitedTo:1e9});case 5:return a.next=7,h.search("",{hitsPerPage:1e9});case 7:return i=a.sent,j=i.hits,a.abrupt("return",j);case 10:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}(),saveItems=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(b,c){var d,e,f,g,h,i,j,k,l,m,n;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return d=c.id,e=c.key,f=c.name,g=void 0===f?"message":f,h=5,i=ora("Connecting to Algolia").start(),j=algoliasearch(d,e),k=j.initIndex(g),a.prev=5,i.text="Saving ".concat(b.length," items..."),l=function(a){return partitionByKeyLength("text",MAX_MESSAGE_TEXT_LENGTH,a)},m=b.flatMap(l).map(function(a,b){return _objectSpread({},a,{objectID:"".concat(a.objectID,"_").concat((b+"").padStart(h,"0"))})}),a.next=11,k.saveObjects(m);case 11:return n=a.sent,i.succeed("Successfully saved ".concat(b.length," (").concat(m.length," total) items!\n")),a.abrupt("return",n);case 16:a.prev=16,a.t0=a["catch"](5),i.fail("Failed to save ".concat(b.length," items")),log.error(a.t0);case 20:case"end":return a.stop();}},a,null,[[5,16]])}));return function(){return a.apply(this,arguments)}}(),populate=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(b,c,d){var e,f,g,h,i;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return e=d.id,f=d.key,g=d.name,a.next=3,getItems(b,c);case 3:return h=a.sent,a.next=6,saveItems(h,{id:e,key:f,name:g});case 6:return i=a.sent,a.abrupt("return",i);case 8:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}(),update=/*#__PURE__*/function(){var a=(0,_asyncToGenerator2["default"])(/*#__PURE__*/_regenerator["default"].mark(function a(b,c){var d,e,f,g,h,i,j,k,l,m,n,o,p,q;return _regenerator["default"].wrap(function(a){for(;;)switch(a.prev=a.next){case 0:return d=c.id,e=c.key,f=c.name,g=c.verbose,h=void 0===g||g,a.next=3,Promise.all([getItems(b,[getCurrentYear()],{verbose:h}),getSavedItems({id:d,key:e,name:f})]);case 3:return i=a.sent,j=(0,_slicedToArray2["default"])(i,2),k=j[0],l=j[1],m=k.map(function(a){var b=a.id;return b}),n=(0,_toConsumableArray2["default"])(new Set(l.map(function(a){var b=a.id;return b}))),o=difference(m,n).sort().map(parseMessageId),a.next=12,Promise.all(o.map(getItem));case 12:if(p=a.sent,q=0===p.length,q&&process.stdout.write("".concat(bold("No records to update"),"\n\n")),!q){a.next=19;break}a.t0={objectIDs:[]},a.next=22;break;case 19:return a.next=21,saveItems(p,{id:d,key:e,name:f});case 21:a.t0=a.sent;case 22:return a.abrupt("return",a.t0);case 23:case"end":return a.stop();}},a)}));return function(){return a.apply(this,arguments)}}();module.exports={getItem:getItem,getItems:getItems,getSavedItems:getSavedItems,populate:populate,saveItems:saveItems,scrapeMessageItems:scrapeMessageItems,update:update};