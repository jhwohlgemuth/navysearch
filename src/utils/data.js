const {promisify} = require('util');
const log = require('npmlog');
const chalk = require('chalk');
const ora = require('ora');
const Xray = require('x-ray');
const algoliasearch = require('algoliasearch');
const axios = require('axios').default;
const {difference} = require('ramda');
const {
    MAX_MESSAGE_TEXT_LENGTH,
    createMessageId,
    createNpcPageUrl,
    createYearsString,
    getCurrentYear,
    parseMessageId,
    parseMessageUri,
    partitionByKeyLength
} = require('./common.js');
const {bold, cyan, green} = chalk;

/**
 * Scrape message data from  NPC
 * @param {string} type NAVADMIN | ALNAV
 * @param {(string|number)} year Last two digits of year of page to scrape from
 * @returns {object[]}
 */
const scrapeMessageItems = async (type, year) => {
    const x = new Xray();
    const url = createNpcPageUrl({type, year});
    const {data} = await axios.get(url);
    const parsedHtml = await promisify(x(data, 'a', [{href: '@href'}]));
    return parsedHtml
        .map(({href}) => href)
        .filter(str => /[.]txt$/.test(str))
        .map(parseMessageUri)
        .filter(({code, num}) => [code, num].every(({length}) => length > 0));
};
const getItem = ({code, num, type, year, url}) => {
    const attributes = {code, num, type, url, year};
    const id = createMessageId(attributes);
    return axios.get(url)
        .then(({data: text}) => ({...attributes, id, text, objectID: id}))
        .catch(() => ({...attributes, id, objectID: id}));
};
const getItems = async (type, years) => {
    const startMessage = items => {
        const [{type}] = items;
        const yearsString = createYearsString(years);
        const message = `Fetching ${bold(type)} data for ${bold(yearsString)} (${items.length} items)\n\n`;
        return cyan(message);
    };
    const doneMessage = items => {
        const completed = items.filter(({text}) => typeof text !== 'undefined');
        return `${green('COMPLETE')} ~ ${bold(completed.length)} of ${bold(items.length)} items processed\n`;
    };
    const get = data => Promise.all(data.map(getItem));
    const items = (await Promise.all(years.map(year => scrapeMessageItems(type, year)))).flat(Infinity);
    const continueOperation = async () => {
        const spinner = ora(startMessage(items)).start();
        const data = await get(items);
        const retry = async data => {
            const failed = data.filter(({text}) => typeof text === 'undefined');
            spinner.text = cyan(`Trying to fetch data for ${bold(failed.length)} of ${items.length} items`);
            return (failed.length !== 0) ? get(failed) : Promise.resolve([]);
        };
        const secondAttempt = await retry(data);
        const thirdAttempt = await retry(secondAttempt);
        const fourthAttempt = await retry(thirdAttempt);
        const results = [
            ...data,
            ...secondAttempt,
            ...thirdAttempt,
            ...fourthAttempt
        ].filter(({text}) => typeof text !== 'undefined');
        spinner.succeed(doneMessage(results));
        return results;
    };
    return items.length > 0 ? continueOperation() : [];
};
const getSavedItems = async ({id, key, name = 'message'}) => {
    const client = algoliasearch(id, key);
    const index = client.initIndex(name);
    await index.setSettings({paginationLimitedTo: 1e9});
    const {hits} = await index.search('', {hitsPerPage: 1e9});
    return hits;
};
/**
 * Save items to Algolia search service
 * @param {object[]} items
 * @param {object} options Configuration options
 * @param {string} options.id Application ID
 * @param {string} options.key Application Admin API key
 * @param {string} [options.name='message'] Index name
 * @returns {object} Results of save operation
 */
const saveItems = async (items, {id, key, name = 'message'}) => {
    const SUFFIX_LENGTH = 5;
    const spinner = ora('Connecting to Algolia').start();
    const client = algoliasearch(id, key);
    const index = client.initIndex(name);
    try {
        spinner.text = `Saving ${items.length} items...`;
        const chunk = value => partitionByKeyLength('text', MAX_MESSAGE_TEXT_LENGTH, value);
        const itemsToUpload = items
            .flatMap(chunk)
            .map((item, index) => ({...item, objectID: `${item.objectID}_${String(index).padStart(SUFFIX_LENGTH, '0')}`}));
        const results = await index.saveObjects(itemsToUpload);
        spinner.succeed(`Successfully saved ${items.length} (${itemsToUpload.length} total) items!\n`);
        return results;
    } catch (error) {
        spinner.fail(`Failed to save ${items.length} items`);
        log.error(error);
    }
};
const populate = async (type, years, {id, key, name}) => {
    const items = await getItems(type, years);
    const results = await saveItems(items, {id, key, name});
    return results;
};
const update = async (type, {id, key, name, verbose = true}) => {
    const [scraped, saved] = await Promise.all([
        getItems(type, [getCurrentYear()], {verbose}),
        getSavedItems({id, key, name})
    ]);
    const left = scraped.map(({id}) => id);
    const right = ([...(new Set(saved.map(({id}) => id)))]);
    const updated = difference(left, right).sort().map(parseMessageId);
    const items = await Promise.all(updated.map(getItem));
    const noItems = items.length === 0;
    if (noItems) {
        process.stdout.write(`${bold('No records to update')}\n\n`);
    }
    return noItems ? {objectIDs: []} : await saveItems(items, {id, key, name});
};

module.exports = {
    getItem,
    getItems,
    getSavedItems,
    populate,
    saveItems,
    scrapeMessageItems,
    update
};