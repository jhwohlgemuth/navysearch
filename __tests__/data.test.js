/**
 * @jest-environment node
 */
import moxios from 'moxios';
import {NPC_DOMAIN, createNpcPageUrl} from '../src/utils/common.js';
import {getItem, getItems, scrapeMessageItems} from '../src/utils/data.js';

const OK = 200;
const NOT_FOUND = 404;
const URL_REGEX = /http[s]?[:]\/\/www[.]public[.]navy[.]mil\/bupers[-]npc\/reference\/messages.*/;
const TEST_MESSAGE = 'INTENTIONALLY LEFT BLANK FOR TESTING';
const fragment = `${NPC_DOMAIN}/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/`;
const NPC_HTML = `
    <!DOCTYPE html>
    <html>
        <body>
            <a href="${fragment}NAV20116.txt"/>
            <a href="/someLocalPath"/>
            <a href="https://civilian.com/some.txt"/>
            <a href="http://${NPC_DOMAIN}/someAbsolutePath"/>
            <a href="${fragment}NAV20117.txt"/>
            <a href="http://${NPC_DOMAIN}/instruction.txt"/>
            <a href="http://${NPC_DOMAIN}/policy.pdf"/>
            <a href="http://${NPC_DOMAIN}/military.txt"/>
            <a href="#"/>
            <a href="miltxt"/>
            <a href="javascript:;"/>
            <a href="${fragment}NAV20118.txt"/>
        </body>
    </html>
`;

describe('Data Utilities', () => {
    beforeEach(() => {
        moxios.install();
    });
    afterEach(() => {
        moxios.uninstall();
    });
    test('can get message item data with getItem', async () => {
        moxios.stubRequest(URL_REGEX, {
            status: OK,
            responseText: TEST_MESSAGE
        });
        const url = `${NPC_DOMAIN}/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20119.txt`;
        const options = {
            url,
            code: 'NAV',
            num: '042',
            type: 'NAVADMIN',
            year: '20'
        };
        expect(await getItem(options)).toMatchSnapshot();
        moxios.stubRequest(URL_REGEX, {
            status: NOT_FOUND,
            responseText: TEST_MESSAGE
        });
    });
    test('can get message data with getItem (failed request)', async () => {
        moxios.stubRequest(URL_REGEX, {
            status: NOT_FOUND
        });
        const url = `${NPC_DOMAIN}/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20119.txt`;
        const options = {
            url,
            code: 'NAV',
            num: '042',
            type: 'NAVADMIN',
            year: '20'
        };
        expect(await getItem(options)).toMatchSnapshot();
    });
    test('can scrape message data from NPC', async () => {
        const type = 'NAVADMIN';
        const year = '20';
        moxios.stubRequest(createNpcPageUrl({type, year}), {
            status: OK,
            responseText: NPC_HTML
        });
        const items = await scrapeMessageItems(type, year);
        expect(items).toMatchSnapshot();
    });
    test('can get data for all messages for a given year on NPC', async () => {
        const type = 'NAVADMIN';
        const year = '20';
        moxios.stubRequest(createNpcPageUrl({type, year}), {
            status: OK,
            responseText: NPC_HTML
        });
        moxios.stubRequest(URL_REGEX, {
            status: OK,
            responseText: TEST_MESSAGE
        });
        const items = await getItems(type, [year]);
        expect(items).toMatchSnapshot();
    });
});