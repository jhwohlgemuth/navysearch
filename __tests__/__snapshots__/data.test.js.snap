// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Utilities can get data for all messages for a given year on NPC 1`] = `
Array [
  Object {
    "code": "NAV",
    "id": "NAVADMIN20116",
    "num": "116",
    "objectID": "NAVADMIN20116",
    "text": "INTENTIONALLY LEFT BLANK FOR TESTING",
    "type": "NAVADMIN",
    "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20116.txt",
    "year": "20",
  },
  Object {
    "code": "NAV",
    "id": "NAVADMIN20117",
    "num": "117",
    "objectID": "NAVADMIN20117",
    "text": "INTENTIONALLY LEFT BLANK FOR TESTING",
    "type": "NAVADMIN",
    "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20117.txt",
    "year": "20",
  },
  Object {
    "code": "NAV",
    "id": "NAVADMIN20118",
    "num": "118",
    "objectID": "NAVADMIN20118",
    "text": "INTENTIONALLY LEFT BLANK FOR TESTING",
    "type": "NAVADMIN",
    "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20118.txt",
    "year": "20",
  },
]
`;

exports[`Data Utilities can get message data with getItem (failed request) 1`] = `
Object {
  "code": "NAV",
  "id": "NAVADMIN20042",
  "num": "042",
  "objectID": "NAVADMIN20042",
  "type": "NAVADMIN",
  "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20119.txt",
  "year": "20",
}
`;

exports[`Data Utilities can get message item data with getItem 1`] = `
Object {
  "code": "NAV",
  "id": "NAVADMIN20042",
  "num": "042",
  "objectID": "NAVADMIN20042",
  "text": "INTENTIONALLY LEFT BLANK FOR TESTING",
  "type": "NAVADMIN",
  "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20119.txt",
  "year": "20",
}
`;

exports[`Data Utilities can scrape message data from NPC 1`] = `
Array [
  Object {
    "code": "NAV",
    "ext": ".txt",
    "id": "NAVADMIN20116",
    "num": "116",
    "type": "NAVADMIN",
    "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20116.txt",
    "year": "20",
  },
  Object {
    "code": "NAV",
    "ext": ".txt",
    "id": "NAVADMIN20117",
    "num": "117",
    "type": "NAVADMIN",
    "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20117.txt",
    "year": "20",
  },
  Object {
    "code": "NAV",
    "ext": ".txt",
    "id": "NAVADMIN20118",
    "num": "118",
    "type": "NAVADMIN",
    "url": "http://www.public.navy.mil/bupers-npc/reference/messages/Documents/NAVADMINS/NAV2020/NAV20118.txt",
    "year": "20",
  },
]
`;
